File: Medium/1577. Number of Ways Where Square of Number Is Equal to Product of Two Numbers/1577. Number of Ways Where Square of Number Is Equal to Product of Two Numbers.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int numTriplets(int[] nums1, int[] nums2) {
        int ans = 0;
        
        Map<Long, Integer> nums1Product = new HashMap<>();
        Map<Long, Integer> nums2Product = new HashMap<>();
        
        // Calculate products of pairs of elements in nums1 and store their counts in nums1Product
        for (int i = 0; i < nums1.length; i++) {
            for (int j = i + 1; j < nums1.length; j++) {
                long product = (long) nums1[i] * nums1[j];
                nums1Product.put(product, nums1Product.getOrDefault(product, 0) + 1);
            }
        }
        
        // Calculate products of pairs of elements in nums2 and store their counts in nums2Product
        for (int i = 0; i < nums2.length; i++) {
            for (int j = i + 1; j < nums2.length; j++) {
                long product = (long) nums2[i] * nums2[j];
                nums2Product.put(product, nums2Product.getOrDefault(product, 0) + 1);
            }
        }
        
        // Check for square numbers in nums1 and add the corresponding product counts from nums2
        for (int num : nums1) {
            long target = (long) num * num;
            ans += nums2Product.getOrDefault(target, 0);
        }
        
        // Check for square numbers in nums2 and add the corresponding product counts from nums1
        for (int num : nums2) {
            long target = (long) num * num;
            ans += nums1Product.getOrDefault(target, 0);
        }
        
        return ans;
    }
}

File: Medium/1577. Number of Ways Where Square of Number Is Equal to Product of Two Numbers/1577. Number of Ways Where Square of Number Is Equal to Product of Two Numbers.rs
impl Solution {
    pub fn num_triplets(nums1: Vec<i32>, nums2: Vec<i32>) -> i32 {
        let mut ans = 0;
        let mut nums1_product = std::collections::HashMap::new();
        let mut nums2_product = std::collections::HashMap::new();

        for (nums, products) in vec![(&nums1, &mut nums1_product), (&nums2, &mut nums2_product)] {
            for i in 0..nums.len() {
                for j in i + 1..nums.len() {
                    let product = (nums[i] as i64) * (nums[j] as i64);
                    *products.entry(product).or_insert(0) += 1;
                }
            }
        }

        for num in &nums1 {
            let target = (*num as i64).pow(2);
            if let Some(&value) = nums2_product.get(&target) {
                ans += value;
            }
        }

        for num in &nums2 {
            let target = (*num as i64).pow(2);
            if let Some(&value) = nums1_product.get(&target) {
                ans += value;
            }
        }

        ans
    }
}
